packages/ra-core/src/i18n/TranslationProvider.tsx

```diff
@@ -56,7 +56,7 @@ class TranslationProviderView extends Component<ViewProps, State> {
     }
 
     componentDidUpdate(prevProps) {
-        if (prevProps.locale !== this.props.locale) {
+        if (prevProps.locale !== this.props.locale || prevProps.messages !== this.props.messages) {
```

Comment:
You're comparing object references here for messages. This should be enough as the reference to the messages object should change when loading new ones